include/swift/SIL/SILType.h

```diff
@@ -553,33 +547,346 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, SILType T) {
   return OS;
 }
 
-inline SILType SILParameterInfo::getSILType() const {
-  if (isIndirect()) {
-    return SILType::getPrimitiveAddressType(getType());
-  } else {
-    return SILType::getPrimitiveObjectType(getType());
+/// Transient wrapper for SILParameterInfo and SILResultInfo conventions. This
+/// abstraction helps handle the transition from canonical SIL conventions to
+/// lowered SIL conventions.
+class SILModuleConventions {
+  friend SILParameterInfo;
+  friend SILResultInfo;
+  friend SILFunctionConventions;
+
+  static bool isIndirectSILParam(SILParameterInfo param, bool loweredAddresses);
+
+  static bool isIndirectSILResult(SILResultInfo result, bool loweredAddresses);
+
+  static SILType getSILParamType(SILParameterInfo param, bool loweredAddresses);
+
+  static SILType getSILResultType(SILResultInfo param, bool loweredAddresses);
+
+public:
+  static bool isPassedIndirectlyInSIL(SILType type, SILModule &M);
+
+  static bool isReturnedIndirectlyInSIL(SILType type, SILModule &M);
+
+private:
+  bool loweredAddresses;
+
+public:
+  SILModuleConventions(const SILModule &M);
+
+  SILFunctionConventions getFunctionConventions(CanSILFunctionType funcTy);
+
+  bool isSILIndirect(SILParameterInfo param) const {
+    return isIndirectSILParam(param, loweredAddresses);
+  }
+
+  bool isSILIndirect(SILResultInfo result) const {
+    return isIndirectSILResult(result, loweredAddresses);
+  }
+
+  SILType getSILType(SILParameterInfo param) const {
+    return getSILParamType(param, loweredAddresses);
+  }
+
+  SILType getSILType(SILResultInfo result) const {
+    return getSILResultType(result, loweredAddresses);
+  }
+};
+
+/// Transient wrapper for SIL-level argument conventions. This abstraction helps
+/// handle the transition from canonical SIL conventions to lowered SIL
+/// conventions.
+class SILFunctionConventions {
+public:
+  SILModuleConventions silConv;
```

Comment:
The "conventions" object can be on-demand. A getter would logically construct
a new one each time. That's not what I want for most uses.

More importantly, `silConv` is an interface to conceptually global
context. Like `M`, `F`, `IGM`, `SGM`, etc. It's important to
standardize abbreviations for these. They should be accessed using
common idioms throughout the compiler.

Defining getters for global context is superfluous. It is not part of
the class's implementation detail. The only reason we attach these
contextual artifacts to most of our classes is because of a decision
not to use TLS within the compiler. That decision makes exploratory
programming hard, and makes it difficult to establish uniform
idioms. We can improve the situation by standarizing on simple
abbreviated names for contextual state and making them a public member
of every class that accesses them.

I'm taking the time to explain this not because of some theoretical
concern but because lack of standard idioms is a real problem for me
when working across different areas of our code base. Most of my
cognitive load goes to figuring out how the same familiar things are
accessed differently in each corner of the code.